### DISCUSSION TITLE
Matching overlapping pattern with DependencyMatcher

### DISCUSSION THREAD

**Discussion by ghost** [explosion/spaCy#11878] [Help: Model Advice]
Here

<img width="1016" alt="Screenshot 2022-11-26 at 13 11 00" src="https://user-images.githubusercontent.com/48595253/204083481-6acde9cf-7242-4eca-b573-61d533801280.png">

I would expect to get 2 matches for "by" and for "to" but I only receive the first one ("by") Does DependencyMatcher allows matching "overlapping" patterns (ie the VERB is the same in both matches)

```python
import spacy 
from spacy.matcher import DependencyMatcher

nlp = spacy.load('en_core_web_lg')

matcher = DependencyMatcher(nlp.vocab)
pattern = [
  {
    "RIGHT_ID": "anchor_VERB",
    "RIGHT_ATTRS": {"POS": {"IN": ["VERB"]}},
  },
  {
    "LEFT_ID": "anchor_VERB",
    "REL_OP": ">",
    "RIGHT_ID": "anchor_POS", 
    "RIGHT_ATTRS": {"DEP": "prep"}
  },
]
matcher.add("BY", [pattern])

doc = nlp("We increased revenue by $1.7 billion to $3.9 billion.")

matches = matcher(doc)

print(matches)

# Each token_id corresponds to one pattern dict
for match_id, token_ids in matches:
  for i in range(len(token_ids)):
    tok = doc[token_ids[i]]
    print(f'''{pattern[i]["RIGHT_ID"]}: "{tok.text}" ''')
  print('\n')
```

Output:

```
[(5422785675429246524, [1, 3])]
anchor_VERB: "increased" 
anchor_POS: "by" 
```

**Comment by ghost** [comment]
And this is weird (a bug?) 

I'm using `>` which is *immediate head* but it picks leaves from different sub-trees for "by" and "to"

<img width="1015" alt="Screenshot 2022-11-26 at 13 38 06" src="https://user-images.githubusercontent.com/48595253/204084533-457cb903-932b-4978-843b-3224ac774789.png">


```python
import spacy 
from spacy.matcher import DependencyMatcher

nlp = spacy.load('en_core_web_lg')

matcher = DependencyMatcher(nlp.vocab)
pattern = [
  {
    "RIGHT_ID": "anchor_POS",
    "RIGHT_ATTRS": {"POS": {"IN": ["ADP", "PART"]}},
  },
  {
    "LEFT_ID": "anchor_POS",
    "REL_OP": ">",
    "RIGHT_ID": "by_howmuch",
    "RIGHT_ATTRS": {"DEP": "pobj"},
  },
  {
    "LEFT_ID": "by_howmuch",
    "REL_OP": ">",
    "RIGHT_ID": "by_howmuch_modifier",
    "RIGHT_ATTRS": {"DEP": {"IN": ["amod", "compound"]}},
  },
]
matcher.add("BY", [pattern])

doc = nlp("We increased revenue by $1.7 billion to $3.9 million.")

matches = matcher(doc)

print(matches)

for match_id, token_ids in matches:
  for i in range(len(token_ids)):
    tok = doc[token_ids[i]]
    print(f'''{pattern[i]["RIGHT_ID"]}: "{tok.text}" ''')
  print('\n')
```

Output:

```[(5422785675429246524, [3, 10, 5]), (5422785675429246524, [3, 10, 9])]
anchor_POS: "by" 
by_howmuch: "million" 
by_howmuch_modifier: "1.7" 


anchor_POS: "by" 
by_howmuch: "million" 
by_howmuch_modifier: "3.9" 
```

**Comment by adrianeboyd** [comment]
I think the problem is that the parse you're getting with the exact version of `en_core_web_lg` used in your script isn't the same as the parse in the screenshot.

Dealing with differences/errors in the dependency parses for difficult cases is one of the tricky parts of writing dependency matcher patterns. For the case in your top post, the dependency matcher should return both matches for the parse in the screenshot, but I think the script isn't analyzing the same parse.

Can you check the parses in your scripts directly with displacy (or by checking the dependency parse with `token.head` and `token.dep_`)?

For example, this is the parse I get with `en_core_web_lg` v3.4.1 (v3.4.1 is the model package version, not the spacy version):
![Screenshot_2022-11-28_08-12-00](https://user-images.githubusercontent.com/5794899/204216057-a393cff7-ac6d-4498-a12e-729efa905667.png)
